File: ios/Classes/PlatoformRequester/AmazonChimeError.swift
//
//  AmazonChimeError.swift
//  amazon_chime_plugin
//
//  Created by Yugo Sugiyama on 2023/08/31.
//

import Foundation
import Flutter

extension FlutterError: Error {}

enum AmazonChimeError: Error {
    case invalidResponse
    case responseMessage(type: ResponseMessageKind)
    case customError(text: String)

    private var errorType: AmazonChimeErrorType {
        switch self {
        case .invalidResponse:
            return .invalidResponse
        case .responseMessage:
            return .responseMessage
        case .customError:
            return .customError
        }
    }

    var asFlutterError: FlutterError {
        let code = String(errorType.rawValue)
        switch self {
        case .invalidResponse:
            return FlutterError(code: code, message: nil, details: nil)
        case .responseMessage(let type):
            return FlutterError(code: code, message: type.rawValue, details: nil)
        case .customError(let text):
            return FlutterError(code: code, message: text, details: nil)
        }
    }
}

File: android/src/main/kotlin/com/example/amazon_chime_plugin/PlatoformRequester/Errors/AmazonChimeError.kt
package com.example.amazon_chime_plugin.PlatoformRequester.Errors
import AmazonChimeErrorType
import FlutterError
import com.example.amazon_chime_plugin.PlatoformRequester.Models.ResponseMessageKind

sealed class AmazonChimeError: Throwable() {
    abstract val errorType: AmazonChimeErrorType
    object InvalidResponse: AmazonChimeError() {
        override val errorType: AmazonChimeErrorType =
            AmazonChimeErrorType.INVALIDRESPONSE
    }
    data class ResponseMessage(val type: ResponseMessageKind): AmazonChimeError() {
        override val errorType: AmazonChimeErrorType =
            AmazonChimeErrorType.RESPONSEMESSAGE
    }
    data class CustomError(val text: String): AmazonChimeError() {
        override val errorType: AmazonChimeErrorType =
            AmazonChimeErrorType.INVALIDRESPONSE
    }

    val asFlutterError: FlutterError
        get() {
            val code = errorType.raw.toString()
            return when (this) {
                is InvalidResponse -> FlutterError(code, null, null)
                is ResponseMessage -> FlutterError(code, type.rawValue, null)
                is CustomError -> FlutterError(code, text, null)
            }
        }
}
